packages/coverage-v8/src/provider.ts

```diff
@@ -457,39 +457,44 @@ export class V8CoverageProvider extends BaseCoverageProvider implements Coverage
 
     const map = transformResult?.map as EncodedSourceMap | undefined
     const code = transformResult?.code
-    const sourcesContent
-      = map?.sourcesContent?.[0]
-      || (await fs.readFile(filePath, 'utf-8').catch(() => {
+    const sourcesContent = map?.sourcesContent || []
+
+    if (!sourcesContent[0]) {
+      sourcesContent[0] = await fs.readFile(filePath, 'utf-8').catch(() => {
         // If file does not exist construct a dummy source for it.
         // These can be files that were generated dynamically during the test run and were removed after it.
         const length = findLongestFunctionLength(functions)
         return '.'.repeat(length)
-      }))
+      })
+    }
 
     // These can be uncovered files included by "all: true" or files that are loaded outside vite-node
     if (!map) {
       return {
         isExecuted,
-        source: code || sourcesContent,
-        originalSource: sourcesContent,
+        source: code || sourcesContent[0],
+        originalSource: sourcesContent[0],
       }
     }
 
-    const sources = [url]
-    if (map.sources && map.sources[0] && !url.endsWith(map.sources[0])) {
-      sources[0] = new URL(map.sources[0], url).href
+    const sources = (map.sources || [])
+      .filter((source): source is NonNullable<typeof source> => !!source)
```

Comment:
you can do `source != null` and you don't even need to have `source is` then :P

this was added in the latest TS version

Reply:
That works, nice! I though it would detect `!!` as well, and thought we were using older TS version. 💯

Reply:
Yeah, unfortunately it doesn't work with `!!` because it's not an explicit assertion 😢